### PR TITLE
Use weighted algorithm for string matching when finding files in repo

### PR DESCRIPTION
**Pull request by wxiaoguang**
This PR is for:
* https://github.com/go-gitea/gitea/issues/20231

Now, when a user searches `word`, they always see `/{word}.txt` before `/{w}e-g{o}t-{r}esult.{d}at`

Demo:

When searching "a", "a.ext" comes first. 

Then when searching "at", the longer matched "template" comes first.

<details>

image

image

</details>

This PR also makes the frontend tests could import feature JS files by introducing `jestSetup.js`

### CODE REVIEW

File: web_src/js/features/repo-findfile.js [hunk 1]
```diff
@@ -1,45 +1,96 @@
...
+    if (res.length % 2 === 0) res.push('');
+    res[res.length - 1] += full.substring(j);
+  }
+  return res;
+}
+
+export function calcMatchedWeight(matchResult) {
+  let weight = 0;
+  for (let i = 0; i < matchResult.length; i++) {
+    if (i % 2 === 1) {
```
Comment (delvh) [inline web_src/js/features/repo-findfile.js]:
```suggestion
    if (i % 2 === 1) { // matches are on odd indices, see strSubMatch
```

File: web_src/js/features/repo-findfile.js [hunk 2]
```diff
@@ -1,45 +1,96 @@
...
-    for (let i = 0; i < files.length && fileRes.length < threshold; i++) {
-      const subMatch = strSubMatch(files[i], filter);
-      if (subMatch.length > 1) {
-        fileRes.push(subMatch);
+    const filterLower = filter.toLowerCase();
+    // TODO: for large repo, this loop could be slow, maybe there could be one more limit:
+    // ... && filterResult.length < threshold * 20,  wait for more feedbacks
+    for (let i = 0; i < files.length; i++) {
+      const res = strSubMatch(files[i], filterLower);
+      if (res.length > 1) {
```
Comment (delvh) [inline web_src/js/features/repo-findfile.js]:
```suggestion
      if (res.length > 1) { // length==1 means unmatched, >1 means having matched sub strings
```

File: web_src/js/features/repo-findfile.js [hunk 3]
```diff
@@ -1,45 +1,96 @@
...
+    if (subLower[i] === fullLower[j]) {
+      if (res.length % 2 !== 0) res.push('');
+      res[res.length - 1] += full[j];
+      j++;
+      i++;
+    } else {
+      if (res.length % 2 === 0) res.push('');
+      res[res.length - 1] += full[j];
+      j++;
+    }
```
Comment (delvh) [inline web_src/js/features/repo-findfile.js:28]:
```suggestion
    if (res.length % 2 !== 0) res.push('');
    res[res.length - 1] += full[j];
    if (subLower[i] === fullLower[j]) {
      i++;
    }
    j++;
```

Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js:28]: The changed code is incorrect. And the old code is also readable.

File: jest.config.js [hunk 4]
```diff
@@ -2,11 +2,13 @@ export default {
   rootDir: 'web_src',
   setupFilesAfterEnv: ['jest-extended/all'],
   testEnvironment: 'jest-environment-jsdom',
-  testMatch: ['<rootDir>/**/*.test.js'],
+  testMatch: ['<rootDir>/**/*.test.js'], // to run tests with ES6 module, node must run with "--experimental-vm-modules"
```
Comment (silverwind) [inline jest.config.js]: We already have this flag in the Makefile, not sure the benefit of putting this comment in this odd place here. Maybe at the start of the file, if at all.

BTW, personally, I just have this in my shell startup file to always run jest without issues:

````bash
export NODE_OPTIONS="--experimental-vm-modules --no-warnings --trace-uncaught"

File: web_src/js/features/repo-findfile.js (hunk 3 above)
Comment (delvh) [reply web_src/js/features/repo-findfile.js:28]: Yep, I just noticed that there is a single difference: `res.length % 2 (!|=)== 0`…

File: web_src/js/features/repo-findfile.test.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import {strSubMatch, calcMatchedWeight, filterRepoFilesWeighted} from './repo-findfile.js';
+import {expect} from '@playwright/test';
```
Comment (silverwind) [inline web_src/js/features/repo-findfile.test.js]:
```suggestion
```

Use the jest global instead. playwright `expect` should only be used in playwright tests.

File: jest.config.js [hunk 6]
```diff
@@ -2,11 +2,13 @@ export default {
...
   testTimeout: 20000,
   transform: {
     '\\.svg$': '<rootDir>/js/testUtils/jestRawLoader.js',
   },
+  setupFiles: [
+    './js/testUtils/jestSetup.js', // prepare global variables used by our code (eg: window.config)
+  ],
   verbose: false,
 };
-
```
Comment (silverwind) [inline jest.config.js:12]: No need to drop final newline.

File: web_src/js/features/repo-findfile.js [hunk 7]
```diff
@@ -1,45 +1,96 @@
...
-function filterRepoFiles(filter) {
-  const treeLink = $repoFindFileInput.attr('data-url-tree-link');
-  $repoFindFileTableBody.empty();
 
-  const fileRes = [];
+// return the case-insensitive sub-match result as an array:  [unmatched, matched, unmatched, matched, ...]
+// res[even] is unmatched, res[odd] is matched, see unit tests for examples
+// argument subLower must be a lower-cased string.
+export function strSubMatch(full, subLower) {
+  const res = [''];
```
Comment (silverwind) [inline web_src/js/features/repo-findfile.js:15]: Can we do without  empty string results?

File: web_src/js/features/repo-findfile.js [hunk 8]
```diff
@@ -1,45 +1,96 @@
...
   }
+  return filterResult;
+}
+
+function filterRepoFiles(filter) {
+  const treeLink = $repoFindFileInput.attr('data-url-tree-link');
+  $repoFindFileTableBody.empty();
 
+  const filterResult = filterRepoFilesWeighted(files, filter);
   const tmplRow = `<tr><td><a></a></td></tr>`;
```
Comment (silverwind) [inline web_src/js/features/repo-findfile.js:80]: I think you could `$('tr><td><a></a></td></tr>')` here instead and use [`.clone()`](https://api.jquery.com/clone/) below in the loop. Thought, no big preference here.

File: jest.config.js (hunk 4 above)
Comment (wxiaoguang) [reply jest.config.js]: It's for people who don't use or read the Makefile

File: web_src/js/features/repo-findfile.test.js (hunk 5 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.test.js]: Hmm, by mistake.

File: web_src/js/features/repo-findfile.js (hunk 8 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js:80]: No XSS. Everything is escaped correctly. No string is put into the HTML string.

File: web_src/js/features/repo-findfile.js (hunk 7 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js:15]: No. The `res[0]` is for the unmatched string from beginning.

File: jest.config.js (hunk 4 above)
Comment (silverwind) [reply jest.config.js]: Then put it at the start of the file.

File: web_src/js/features/repo-findfile.js (hunk 8 above)
Comment (silverwind) [reply web_src/js/features/repo-findfile.js:80]: Yeah I see it only modifies the href which is safe. So feel free to ignore this suggestion.

File: jest.config.js (hunk 6 above)
Comment (wxiaoguang) [reply jest.config.js:12]: But there were 2 new lines before. My change is correct.

image

Comment (wxiaoguang) [reply jest.config.js:12]:
```
curl --proxy http://127.0.0.1:1080/  https://raw.githubusercontent.com/go-gitea/gitea/main/jest.config.js | hexdump
...
0000120 3b7d 0a0a
```

Comment (silverwind) [reply jest.config.js:12]: Huh, you are right. GitHub UI completely hides this fact that there are two newlines, but I see it when I view the file locally. I wonder why eslint didn't catch it.

File: jest.config.js (hunk 4 above)
Comment (wxiaoguang) [reply jest.config.js]: done

**Review by silverwind — APPROVED** [review]

**Review by delvh — APPROVED** [review]

**Pull request merged**